src/nstime.c

```diff
@@ -152,12 +152,41 @@ nstime_monotonic_impl(void) {
 }
 nstime_monotonic_t *JET_MUTABLE nstime_monotonic = nstime_monotonic_impl;
 
+static void
+nstime_get_realtime(nstime_t *time) {
+  struct timespec ts;
+
+  clock_gettime(CLOCK_REALTIME, &ts);
+  nstime_init2(time, ts.tv_sec, ts.tv_nsec);
+}
+
+static bool
+nstime_prof_update_impl(nstime_t *time) {
+  nstime_t old_time;
+
+	nstime_copy(&old_time, time);
+
+  if (unlikely(opt_prof_time_resolution == prof_time_resolution_high)) {
+    nstime_get_realtime(time);
+  } else {
+    nstime_get(time);
+  }
+
+	/* Handle non-monotonic clocks. */
```

Comment:
Given that this is prof-only, I sort of think we should allow time to go backwards; for the internal clocks, we need monotonicity for correctness (i.e. we algorithmically need a nondecreasing sequence). But if the user asks for timestamps per some particular clock, we might as well give it to them.

Reply:
I had included the non-monotonic check just to keep the previous behaviour exactly the same. I agree with you completely though in that we (jemalloc) shouldn't be interfering with timestamps for profiling and that should be left up to the consumer. If this monotonic guarantee for profiling isn't documented I'm happy to remove it.

Reply:
@davidtgoldblatt - I just copied nstime_update_impl() here and I just noticed what a strange interface definition it has. A return of 'false' indicates the work was done successfully because their isn;t a possible failure really. A return of 'true' indicates that a clock went backwards and that was corrected! Hmm. It looks like nobody checks the return values anywhere for this interface.

I'm going to not return anything in my next revisiopn so let me know if you prefer nstime_prof_update_impl() to keep the same return type as nstime_update_impl() .

Reply:
> If this monotonic guarantee for profiling isn't documented I'm happy to remove it.

I'd say go ahead and remove it. We only call the `_prof` variants down prof pathways, so it won't affect non-prof logic.

> I'm going to not return anything in my next revision so let me know if you prefer nstime_prof_update_impl() to keep the same return type as nstime_update_impl() .

We might as well just have all of these return void; I think it's probably only historical cruft that makes us not do so.